File: Scraps/java/Coding_Practice/src/leetCode/CopyRandomList.java
package leetCode;

import java.util.HashMap;
import java.util.Map;

public class CopyRandomList {

	class Node {
		public int val;
		public Node next;
		public Node random;

		public Node() {
		}

		public Node(int _val, Node _next, Node _random) {
			val = _val;
			next = _next;
			random = _random;
		}
	};

	public Node copyRandomList(final Node head) {
		final Map<Node, Node> map = new HashMap<>();
		
		if (head == null)
			return null;
		Node cpyHead = new Node(head.val, head.next, head.random);
		map.put(head, cpyHead);
		
		Node curr = head.next;
		Node cpy = cpyHead;
		
		while (true) {
			if (curr == null)
				break;
			final Node cpy2 = new Node(curr.val, curr.next, curr.random);
			cpy.next = cpy2;
			cpy = cpy2;
			map.put(curr, cpy);
			curr = curr.next;
		}

		cpy = cpyHead;
		while (true) {
			if (cpy == null)
				break;
			cpy.random = map.get(cpy.random);
			cpy = cpy.next;
		}
		return cpyHead;
	}
}

File: Scraps/Python/py_scraps/leetcode/CopyRandomList.py
# Definition for a Node.
class Node:
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if head is None:
            return None

        memo = {}
        new_head = Node(head.val, None, None)
        memo[head] = new_head

        curr = head.next
        new_curr = new_head

        while curr is not None:
            new_node = Node(curr.val, None, None)
            new_curr.next = new_node
            new_curr = new_node
            memo[curr] = new_curr
            curr = curr.next

        new_curr = new_head
        curr = head
        while new_curr is not None:
            if curr.random is None:
                new_curr.random = None
            else:
                new_curr.random = memo[curr.random]
            new_curr = new_curr.next
            curr = curr.next

        return new_head
